src/System.Memory/src/System/MemoryExtensions.cs

```diff
@@ -17,10 +17,227 @@ namespace System
     /// </summary>
     public static partial class MemoryExtensions
     {
+        /// <summary>
+        /// Moves the source characters to the destination with a specified number of characters starting at a specified position removed.
+        /// </summary>
+        /// <param name="source">The source span from which the characters are removed.</param>
+        /// <param name="startIndex">The zero-based position to begin deleting characters.</param>
+        /// <param name="count">The number of characters to delete.</param>
+        /// <param name="destination">The span to copy the leftover characters into.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when the destination Span is too small to contain the source Span with the characters removed
+        /// (i.e. destination.Length &lt; source.Length - count).
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when either <paramref name="startIndex"/> or <paramref name="count"/> is less than zero or
+        /// when <paramref name="startIndex"/> plus <paramref name="count"/> specify a position outside the source span.
+        /// </exception>
+        public static void Remove(this ReadOnlySpan<char> source, int startIndex, int count, Span<char> destination)
```

Comment:
Would it make sense for the method to either return the length of the result or require the destination buffer to be exactly the size of the result. As it stands right now, the caller has to compute the size of the result.

Reply:
@KrzysztofCwalina Have we looked at Span libraries in other programming environments and see what they design patterns are? Do they have Remove API like this?

This Remove API does not feel right. I do not see why anybody would ever want to use it instead of just a simple CopyTo that is much easier to understand.

Reply:
I agree with Jan here. It can be confusing too to have it. This API can be a useful APII only if we make it removing some characters patterns from the source span. but with the proposed behavior, CopyTo should be enough and clearer.

Reply:
I agree. I think it would be better for this API to be in-place instead of taking in a destination buffer, especially given we already deviate from the string.Remove signature here.

Reply:
This API is for transitioning existing string code to the world of spans. I think it's not relevant if other languages have it or not. The motivation is that string has these APIs.

But, I agree with removing this particular method. I think it does too much dancing around the overlapping spans. In real scenarios, the caller will know the relative locations of the buffers and will be able to write much simpler code.

Reply:
> This API is for transitioning existing string code to the world of spans

Then we should do a little validation for each of these APIs that they actually help with transitioning existing code to the world of spans. Pick a few real world examples and transition them over to Spans using these APIs.

For example, here is randomly choosen use of `string.Remove` in corefx: https://github.com/dotnet/corefx/blob/master/src/System.Management/src/System/Management/ManagementQuery.cs#L122 . What we want this to look like when transitioned to Spans?

Reply:
This isn't necessarily a good example since all the calls to Remove here can be replaced with Slice (since characters are only being removed from the front). Th real scenario would be something that requires removal from the middle.

In any case, it could end up looking like this, but it won't use Remove (not tested):
```C#
internal static void ParseToken(ref ReadOnlySpan<char> q, ReadOnlySpan<char> token, ReadOnlySpan<char> op, ref bool bTokenFound, ref ReadOnlySpan<char> tokenValue)
{
    if (bTokenFound)
        throw new ArgumentException();	// Invalid query - duplicate token

    bTokenFound = true;
    q = q.Slice(token.Length).TrimStart(null);

    if (0 != q.IndexOf(op))
        throw new ArgumentException();	// Invalid query

    // Strip off the op and any leading WS
    q = q.Slice(op.Length).TrimStart(null);

    if (0 == q.Length)
        throw new ArgumentException();      // Invalid query - token has no value

    // Next token should be the token value - look for terminating WS 
    // or end of string
    int i;
    if (-1 == (i = q.IndexOf(' ')))
        i = q.Length;           // No WS => consume entire string

    tokenValue = q.Slice(0, i);
    q = q.Slice(tokenValue.Length).TrimStart(null);
}
```

If we insist on using Remove instead, maybe something like this:
```C#
internal static string ParseToken(ReadOnlySpan<char> q, ReadOnlySpan<char> token, ReadOnlySpan<char> op, ref bool bTokenFound, out int bytesConsumed)
{
    if (bTokenFound)
        throw new ArgumentException();	// Invalid query - duplicate token

    bTokenFound = true;
    Span<char> result = new char[q.Length];
    q.Remove(0, token.Length, result);
    result.AsReadOnlySpan().TrimStart(null);

    if (0 != result.IndexOf(op))
        throw new ArgumentException();	// Invalid query

    // Strip off the op and any leading WS
    result.AsReadOnlySpan().Remove(0, op.Length, result);
    result.AsReadOnlySpan().TrimStart(null);

    if (0 == result.Length)
        throw new ArgumentException();      // Invalid query - token has no value

    // Next token should be the token value - look for terminating WS 
    // or end of string
    int i;
    if (-1 == (i = result.IndexOf(' ')))
        i = result.Length;           // No WS => consume entire string

    bytesConsumed = result.Slice(i).AsReadOnlySpan().TrimStart(null).Length;
    return new string(result.Slice(0, i).ToArray());
}
```

Reply:
An example where startIndex != 0 is:
https://github.com/dotnet/corefx/blob/master/src/System.Private.Uri/src/System/Uri.cs#L957

But even then, we can work around it:
```C#
                if (NotAny(Flags.HostNotCanonical | Flags.PathNotCanonical | Flags.ShouldBeCompressed))
                {
                    start = IsUncPath ? _info.Offset.Host - 2 : _info.Offset.Path;

                    ReadOnlySpan<char> str = _string.AsReadOnlySpan();
                    if (!(IsImplicitFile && _info.Offset.Host == (IsDosPath ? 0 : 2) && info.Offset.Query == _info.Offset.End))
                    {
                        str = (IsDosPath && (str[start] == '/' || str[start] == '\\'))
                                ? str.Slice(start + 1, _info.Offset.Query - start - 1)
                                : str.Slice(start, _info.Offset.Query - start);
                    }

                    // Should be a rare case, convert c|\ into c:\
                    if (IsDosPath && str[1] == '|')
                    {
                        str[1] = ':';
                    }
```

Reply:
> `str[1] = ':';`

Span is read-only in this example. I think you would want something like this:

```
char[] arr = str.ToArray();
arr[1] = '1';
str = arr.AsReadOnlySpan();
```

Reply:
Yes. You are right.

I will get rid of Remove() as we have it currently then. It doesn't seem necessary and the current API signature just adds complexity.